File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/network/GeminiClient.kt
package com.jworks.kanjiquest.android.network

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

class GeminiClient(private val apiKey: String) {

    private val endpoint =
        "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent"

    suspend fun generateWithImage(prompt: String, imageBase64: String): String =
        withContext(Dispatchers.IO) {
            val url = URL("$endpoint?key=$apiKey")
            val conn = url.openConnection() as HttpURLConnection
            try {
                conn.requestMethod = "POST"
                conn.setRequestProperty("Content-Type", "application/json")
                conn.doOutput = true
                conn.connectTimeout = 10_000
                conn.readTimeout = 30_000

                val body = JSONObject().apply {
                    put("contents", JSONArray().put(
                        JSONObject().put("parts", JSONArray().apply {
                            put(JSONObject().put("text", prompt))
                            put(JSONObject().put("inline_data", JSONObject().apply {
                                put("mime_type", "image/png")
                                put("data", imageBase64)
                            }))
                        })
                    ))
                }

                conn.outputStream.bufferedWriter().use { it.write(body.toString()) }

                if (conn.responseCode !in 200..299) {
                    val errorBody = conn.errorStream?.bufferedReader()?.readText() ?: ""
                    throw GeminiException("HTTP ${conn.responseCode}: $errorBody")
                }

                val response = conn.inputStream.bufferedReader().readText()
                val json = JSONObject(response)

                // Extract text from Gemini response:
                // { candidates: [{ content: { parts: [{ text: "..." }] } }] }
                json.getJSONArray("candidates")
                    .getJSONObject(0)
                    .getJSONObject("content")
                    .getJSONArray("parts")
                    .getJSONObject(0)
                    .getString("text")
            } finally {
                conn.disconnect()
            }
        }

    suspend fun isAvailable(): Boolean = withContext(Dispatchers.IO) {
        apiKey.isNotBlank()
    }
}

class GeminiException(message: String) : Exception(message)

File: ios-app/KanjiQuest/Network/GeminiClient.swift
import Foundation

/// HTTP client for Gemini API calls (writing feedback, AI grading).
/// Mirrors Android's GeminiClient.
class GeminiClient {
    let apiKey: String

    init(apiKey: String) {
        self.apiKey = apiKey
    }

    func generateContent(prompt: String, imageData: Data? = nil) async throws -> String {
        let url = URL(string: "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=\(apiKey)")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        var parts: [[String: Any]] = [["text": prompt]]

        if let imageData = imageData {
            let base64 = imageData.base64EncodedString()
            parts.append([
                "inline_data": [
                    "mime_type": "image/png",
                    "data": base64
                ]
            ])
        }

        let body: [String: Any] = [
            "contents": [["parts": parts]]
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, _) = try await URLSession.shared.data(for: request)
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let candidates = json?["candidates"] as? [[String: Any]]
        let content = candidates?.first?["content"] as? [String: Any]
        let responseParts = content?["parts"] as? [[String: Any]]
        return responseParts?.first?["text"] as? String ?? ""
    }
}
